File: java/src/main/java/com/android/util/OnAppBarStateChangeListener.java
package com.android.util;

import com.google.android.material.appbar.AppBarLayout;

/**
 * Created by xuzhb on 2021/1/27
 * Desc:监听AppBarLayout的展开与折叠
 */
public abstract class OnAppBarStateChangeListener implements AppBarLayout.OnOffsetChangedListener {

    //调用示例
//    appBarLayout.addOnOffsetChangedListener(new OnAppBarStateChangeListener() {
//        @Override
//        public void onStateChanged(AppBarLayout appBarLayout, State state) {
//
//        }
//    });

    public enum State {
        EXPANDED,
        COLLAPSED,
        IDLE
    }

    private State mCurrentState = State.IDLE;

    @Override
    public final void onOffsetChanged(AppBarLayout appBarLayout, int i) {
        if (i == 0) {
            if (mCurrentState != State.EXPANDED) {
                onStateChanged(appBarLayout, State.EXPANDED);
            }
            mCurrentState = State.EXPANDED;
        } else if (Math.abs(i) >= appBarLayout.getTotalScrollRange()) {
            if (mCurrentState != State.COLLAPSED) {
                onStateChanged(appBarLayout, State.COLLAPSED);
            }
            mCurrentState = State.COLLAPSED;
        } else {
            if (mCurrentState != State.IDLE) {
                onStateChanged(appBarLayout, State.IDLE);
            }
            mCurrentState = State.IDLE;
        }
    }

    public abstract void onStateChanged(AppBarLayout appBarLayout, State state);

}

File: kotlin/src/main/java/com/android/util/OnAppBarStateChangeListener.kt
package com.android.util

import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.AppBarLayout.OnOffsetChangedListener
import kotlin.math.abs

/**
 * Created by xuzhb on 2021/1/25
 * Desc:监听AppBarLayout的展开与折叠
 */
abstract class OnAppBarStateChangeListener : OnOffsetChangedListener {

    //调用示例
//    appBarLayout.addOnOffsetChangedListener(object : OnAppBarStateChangeListener() {
//        override fun onStateChanged(appBarLayout: AppBarLayout, state: State) {
//            if (state == State.COLLAPSED) {
//            }
//        }
//    })

    enum class State {
        EXPANDED,   //展开状态
        COLLAPSED,  //折叠状态
        IDLE        //中间状态
    }

    private var mCurrentState = State.IDLE

    override fun onOffsetChanged(appBarLayout: AppBarLayout, i: Int) {
        when {
            i == 0 -> {
                if (mCurrentState != State.EXPANDED) {
                    onStateChanged(appBarLayout, State.EXPANDED)
                }
                mCurrentState = State.EXPANDED
            }
            abs(i) >= appBarLayout.totalScrollRange -> {
                if (mCurrentState != State.COLLAPSED) {
                    onStateChanged(appBarLayout, State.COLLAPSED)
                }
                mCurrentState = State.COLLAPSED
            }
            else -> {
                if (mCurrentState != State.IDLE) {
                    onStateChanged(appBarLayout, State.IDLE)
                }
                mCurrentState = State.IDLE
            }
        }
    }

    abstract fun onStateChanged(appBarLayout: AppBarLayout, state: State)

}